File: src/lib/email-template.ts
import { format } from 'date-fns';
import { RankedEvent, User } from './supabase';

export class EmailTemplateService {
  /**
   * Generate personalized HTML email for a user
   */
  generatePersonalizedEmail(user: User, events: RankedEvent[], feedbackBaseUrl: string): string {
    const eventCards = events
      .map(
        (event) => `
      <div style="background: white; border-radius: 8px; padding: 20px; margin-bottom: 20px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);">
        <h3 style="color: #1a73e8; margin-top: 0;">${this.escapeHtml(event.title)}</h3>
        <p style="color: #666; margin: 10px 0;">
          <strong>When:</strong> ${format(new Date(event.event_date), 'MMMM d, yyyy \'at\' h:mm a')}<br>
          <strong>Where:</strong> ${this.escapeHtml(event.location || 'TBA')}<br>
          <strong>Category:</strong> ${this.formatCategory(event.category || 'general')}
        </p>
        <p style="color: #333; line-height: 1.6;">${this.escapeHtml(event.description || 'No description available')}</p>
        <div style="margin-top: 15px;">
          <span style="background: #e8f0fe; color: #1a73e8; padding: 4px 8px; border-radius: 4px; font-size: 12px; margin-right: 5px;">
            Match: ${Math.round(event.cluster_match * 100)}%
          </span>
          <span style="background: #fef7e0; color: #f9ab00; padding: 4px 8px; border-radius: 4px; font-size: 12px;">
            Score: ${Math.round(event.score)}
          </span>
        </div>
        <div style="margin-top: 15px;">
          <a href="${this.escapeHtml(feedbackBaseUrl)}?uid=${this.escapeHtml(user.id)}&eid=${this.escapeHtml(event.id)}&action=like" 
             style="background: #34a853; color: white; padding: 8px 16px; text-decoration: none; border-radius: 4px; margin-right: 10px; display: inline-block;">
            Interested
          </a>
          <a href="${this.escapeHtml(feedbackBaseUrl)}?uid=${this.escapeHtml(user.id)}&eid=${this.escapeHtml(event.id)}&action=dislike" 
             style="background: #ea4335; color: white; padding: 8px 16px; text-decoration: none; border-radius: 4px; display: inline-block;">
            Not Interested
          </a>
        </div>
      </div>
    `
      )
      .join('');

    return `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Your Personalized Event Newsletter</title>
</head>
<body style="margin: 0; padding: 0; background-color: #f5f5f5; font-family: Arial, sans-serif;">
  <div style="max-width: 600px; margin: 0 auto; background-color: white;">
    <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center;">
      <h1 style="margin: 0; font-size: 28px;">Your Weekly Event Picks</h1>
      <p style="margin: 10px 0 0 0; opacity: 0.9;">Personalized just for you, ${this.escapeHtml(user.name || user.email)}</p>
    </div>
    
    <div style="padding: 30px;">
      <p style="color: #666; line-height: 1.6;">
        Hi ${this.escapeHtml(user.name || 'there')}! 👋<br><br>
        We've curated these ${events.length} events based on your interests and preferences. 
        Click the buttons to let us know what you think!
      </p>
      
      ${eventCards}
      
      <div style="margin-top: 30px; padding: 20px; background: #f8f9fa; border-radius: 8px; text-align: center;">
        <p style="color: #666; margin: 0; font-size: 14px;">
          Want to update your preferences? Reply to this email or visit your dashboard.
        </p>
      </div>
    </div>
    
    <div style="background: #f8f9fa; padding: 20px; text-align: center; color: #999; font-size: 12px;">
      <p style="margin: 0;">
        University Event Newsletter System<br>
        Powered by AI and Machine Learning
      </p>
    </div>
  </div>
</body>
</html>
    `;
  }

  /**
   * Generate batch email template for a cluster
   */
  generateClusterTemplate(clusterEvents: RankedEvent[], clusterId: number): string {
    return `Cluster ${clusterId} Template - ${clusterEvents.length} events`;
  }

  /**
   * Escape HTML to prevent XSS
   */
  private escapeHtml(text: string): string {
    const map: Record<string, string> = {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      '"': '&quot;',
      "'": '&#039;',
    };
    return text.replace(/[&<>"']/g, (m) => map[m]);
  }

  /**
   * Format category name for display
   */
  private formatCategory(category: string): string {
    return category
      .split('_')
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
      .join(' ');
  }
}

File: python/lib/email_template.py
"""
Email Template Service for Connect3

Generates HTML email templates for personalized newsletters
"""
from typing import List, Dict, Any
from datetime import datetime
from jinja2 import Template


# Simple email template
EMAIL_TEMPLATE = """
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Your Event Picks</title>
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 20px; background: #f5f5f5; }
        .container { max-width: 600px; margin: 0 auto; background: white; border-radius: 12px; overflow: hidden; box-shadow: 0 2px 8px rgba(0,0,0,0.1); }
        .header { background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; }
        .header h1 { margin: 0; font-size: 24px; }
        .content { padding: 20px; }
        .event { border: 1px solid #eee; border-radius: 8px; margin-bottom: 16px; overflow: hidden; }
        .event-header { background: #f8f9fa; padding: 12px 16px; border-bottom: 1px solid #eee; }
        .event-title { font-weight: 600; color: #333; margin: 0; }
        .event-category { font-size: 12px; color: #667eea; text-transform: uppercase; }
        .event-body { padding: 16px; }
        .event-desc { color: #666; line-height: 1.5; }
        .event-meta { display: flex; gap: 16px; margin-top: 12px; font-size: 14px; color: #888; }
        .feedback-btns { margin-top: 12px; }
        .btn { display: inline-block; padding: 8px 16px; border-radius: 6px; text-decoration: none; font-size: 14px; margin-right: 8px; }
        .btn-like { background: #28a745; color: white; }
        .btn-dislike { background: #dc3545; color: white; }
        .footer { padding: 20px; text-align: center; color: #888; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🎓 Your Event Picks</h1>
            <p>Hi {{ user_name }}, here are events curated just for you!</p>
        </div>
        <div class="content">
            {% for event in events %}
            <div class="event">
                <div class="event-header">
                    <p class="event-category">{{ event.category or 'Event' }}</p>
                    <h3 class="event-title">{{ event.title }}</h3>
                </div>
                <div class="event-body">
                    <p class="event-desc">{{ event.description[:200] }}{% if event.description|length > 200 %}...{% endif %}</p>
                    <div class="event-meta">
                        {% if event.event_date %}<span>📅 {{ event.event_date[:10] }}</span>{% endif %}
                        {% if event.location %}<span>📍 {{ event.location }}</span>{% endif %}
                    </div>
                    <div class="feedback-btns">
                        <a href="{{ feedback_url }}?user_id={{ user_id }}&event_id={{ event.id }}&action=like" class="btn btn-like">👍 Interested</a>
                        <a href="{{ feedback_url }}?user_id={{ user_id }}&event_id={{ event.id }}&action=dislike" class="btn btn-dislike">👎 Not for me</a>
                    </div>
                </div>
            </div>
            {% endfor %}
        </div>
        <div class="footer">
            <p>Connect3 - Personalized Event Recommendations</p>
            <p>Your feedback helps us improve recommendations!</p>
        </div>
    </div>
</body>
</html>
"""


class EmailTemplateService:
    """Service for generating HTML email templates"""
    
    def __init__(self):
        self.template = Template(EMAIL_TEMPLATE)
    
    def generate_personalized_email(
        self, 
        user: Dict[str, Any], 
        events: List[Dict[str, Any]], 
        feedback_url: str
    ) -> str:
        """Generate personalized HTML email for a user"""
        return self.template.render(
            user_name=user.get("name") or user.get("email", "").split("@")[0],
            user_id=user["id"],
            events=events,
            feedback_url=feedback_url
        )
    
    def generate_text_email(
        self, 
        user: Dict[str, Any], 
        events: List[Dict[str, Any]]
    ) -> str:
        """Generate plain text version of email"""
        user_name = user.get("name") or user.get("email", "").split("@")[0]
        
        lines = [
            f"Hi {user_name}!",
            "",
            "Here are events curated just for you:",
            ""
        ]
        
        for i, event in enumerate(events, 1):
            lines.append(f"{i}. {event.get('title', 'Event')}")
            if event.get("category"):
                lines.append(f"   Category: {event['category']}")
            if event.get("event_date"):
                lines.append(f"   Date: {event['event_date'][:10]}")
            if event.get("location"):
                lines.append(f"   Location: {event['location']}")
            lines.append("")
        
        lines.extend([
            "---",
            "Connect3 - Personalized Event Recommendations"
        ])
        
        return "\n".join(lines)


# Singleton instance
email_template_service = EmailTemplateService()
